tools/test_frontend.py

```diff
@@ -28,6 +25,30 @@
 TEST_CONFIG = 'test_config.yaml'
 
 
+def all_services_running():
```

Comment:
Let's also pop in a docstring or note here.  Otherwise, it looks quite odd that we are checking whether services either are running or exited, and then report success.  It may not be obvious to a reader that there are other exit statuses.